src/ReturnTypes/RequestFileExtension.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * This file is part of Larastan.
+ *
+ * (c) Nuno Maduro <enunomaduro@gmail.com>
+ *
+ *  For the full copyright and license information, please view the LICENSE
+ *  file that was distributed with this source code.
+ */
+
+namespace NunoMaduro\Larastan\ReturnTypes;
+
+use PHPStan\Type\Type;
+use PHPStan\Analyser\Scope;
+use PHPStan\Type\MixedType;
+use Illuminate\Http\Request;
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+
+/**
+ * @internal
+ */
+final class RequestFileExtension implements DynamicMethodReturnTypeExtension
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getClass(): string
+    {
+        return Request::class;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isMethodSupported(MethodReflection $methodReflection): bool
+    {
+        return $methodReflection->getName() === 'file';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTypeFromMethodCall(
+        MethodReflection $methodReflection,
+        MethodCall $methodCall,
+        Scope $scope
+    ): Type {
+        return new MixedType;
```

Comment:
Why mixed? Isn't there a more specific type we can use?

Reply:
Return type from `file()` is `\Illuminate\Http\UploadedFile|\Illuminate\Http\UploadedFile[]|array|null`. As far as I could find in the list of types, this looked like the only fit? Or have i overlooked something.

Reply:
Definitely not - please watch this talk about in-depth look at PHPStan's typesystem: https://www.youtube.com/watch?v=85Aq2rWpGQE

We should differentiate when it returns the object, when it returns the array and when it returns null:

```
new ObjectType(\Illuminate\Http\UploadedFile|\Illuminate\Http\UploadedFile::class)
new ArrayType(new IntegerType(), new ObjectType(\Illuminate\Http\UploadedFile::class))
new NullType()
```

I guess that it returns the object when an argument is present and it returns the array when there's not argument and the method is called like this: `$request->file()`

Not sure when it returns null.

Reply:
I think it returns null when `$default` is not specified (default for `$default` is null) and the `$target` is not a Collection or Array? I will look into the youtube video and update the PR. Thanks!

Reply:
In this scenario, accessing the `file()` method which accesses the `data_get()` method, the `$target` will never be null. It will always be an array. So would this new implementation suffice?

Reply:
Nevermind, it will definitely return null when the `$key` does not exist in the `$target` variable and `$default` is not set to anything but null. How would one go about specifying this as we have no influence over the input array?

Reply:
Can we recognize it statically? Is it defined somewhere that the key must be there? Or should we switch the return types to array|null + object|null?

Reply:
I think we can recognise it statically?

Paths: 
- `$key` input null, returns whole array (which may be empty).
- `$key` existent returns UploadedFile
- `$key` defined but non existent in input array returns `$default` which has a default value of null.

But, how would one check for this? Also, return type can be array + object|null right?

Reply:
Where's the `$default`? It's the second parameter of the method?

Reply:
And how do you find out that `$key` is existent?

Reply:
Yes its the second parameter:
`public function file($key = null, $default = null)` [InteractsWithInput.php](https://github.com/laravel/framework/blob/5.7/src/Illuminate/Http/Concerns/InteractsWithInput.php#L375)

And the `data_get()` method will check that for you [here](https://github.com/laravel/framework/blob/5.7/src/Illuminate/Support/helpers.php#L475).

Reply:
So that's easy - check `$methodCall->args[1]` - if it does not exist, consider it NullType, if it exists, use `$scope->getType($methodCall->args[1]->value)`.

Reply:
> check `$methodCall->args[1]` - if it does not exist, consider it NullType
That is not true, because even if `args[1]` is not defined, it can still return an `UploadedFile`.

Reply:
So actually if `args[1]` is not defined, it is actually a `UploadedFile|null` type.

Reply:
Yes, that's what I was saying. You need to `TypeCombinator::union(new ObjectType(UploadedFile), $defaultType)` and return that.

Reply:
Ah i misunderstood you, now that i've read it twice i fixed it. This gives me an error on the test tho actually portraying the original problem but narrowed down the types. `Cannot call method store() on Illuminate\Http\UploadedFile|null.`. Which would make sense, if you do not verify your request at first? But if you do, and of course static code analysis has no knowledge of that, there wouldn't be a problem.

Reply:
I think this error is fine. It shows up only on level 7 - the strictest level.

Reply:
Well, i guess i should start adding defaults to my project then. Thanks for the support and feedback!

Reply:
Cool - I'm glad there's now one more person that can write PHPStan extensions :)

Reply:
I will definitely keep an eye on the issues here :)